PL2/PL2-core/src/main/java/nl/tudelft/pl2016gr2/core/algorithms/FilterBubbles.java

```diff
@@ -0,0 +1,149 @@
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.PriorityQueue;
+import java.util.Queue;
+
+/**
+ * Class used to filter bubbles based on the phylogenetic tree. 
+ * @author Casper
+ *
+ */
+public class FilterBubbles {
+  
+  //private GraphInterface graph;
+  private OriginalGraph originalGraph;
+  private IPhylogeneticTreeNode treeRoot;
+  private HashSet<Integer> allSharedNodes = new HashSet<>();
+  private int mutationId;
+  
+  /**
+   * Creates a FilterBubbles object, with the graph to be filtered and the node of the phylogenetic
+   * tree based on which this graph will be filtered.
+   * @param originalGraph : the graph to be filtered.
+   * @param root : the node of the phylogenetic tree to use for filtering. This will be the root of 
+   *     the tree when a full overview is wanted.
+   */
+  public FilterBubbles(OriginalGraph originalGraph, IPhylogeneticTreeNode root) {
+    //this.graph = graph;
+    this.originalGraph = originalGraph;
+    this.treeRoot = root;
+    mutationId = 10000;
+  }
+  
+  /**
+   * Filters the input graph of this object, using the phylogenetic tree. It finds all leaves of the
+   * current node of the tree, which correspond to a certain genome. The filtered graph then only 
+   * shows nodes that have all these genomes going through them, and shows mutation nodes in between
+   * these.
+   * @return : a filtered graph.
+   */
+  public GraphInterface filter() {
+    GraphInterface filteredGraph = new BubbledGraph();
+    
+    PriorityQueue<Integer> sharedNodes = getSharedNodes();
+    buildGraph(filteredGraph, sharedNodes);
+    
+    return filteredGraph;
+  }
+  
+  /**
+   * Builds the filtered graph. It loops through the shared nodes (which are ordered on id),
+   * and adds each shared node to the filtered graph. Between two subsequent shared nodes,
+   * a bubble is placed. 
+   * @param filteredGraph : the resulting filtered graph.
+   * @param sharedNodes : a list of shared nodes, sorted on id.
+   */
+  private void buildGraph(GraphInterface filteredGraph, PriorityQueue<Integer> sharedNodes) {
+    Iterator<Integer> it = sharedNodes.iterator();
+    int previous = it.next();
+    
+    while (it.hasNext()) {
+      Node node = originalGraph.getNode(previous);
+      node.getOutlinks().clear();
+      node.getInlinks().clear();
+      node.addOutlink(mutationId);
+      
+      int next = it.next();
+      Node end = originalGraph.getNode(next);
+      end.getInlinks().clear();
+      end.getOutlinks().clear();
+      end.addInlink(mutationId);
+      
+      // Fix sequencelength
+      Bubble bubble = new Bubble(mutationId, 1);
+      bubble.addInlink(node.getId());
+      bubble.addOutlink(end.getId());
+      
+      filteredGraph.addNode(node);
+      filteredGraph.addNode(bubble);
+      filteredGraph.addNode(end);
+      
+      previous = next;
+      mutationId++;
+    }
```

Comment:
FilterBubbles.java contains many long methods. This one will trigger the methodLength checkstyle rule.
Consider creating private methods that deal with a single block, or increasing functionality of the model classes (`Bubble` / `Node`) if there is some modifications that should be handled by the class itself.

`end.getInlinks().clear();` (line `80`) is an example of using a getter to modify the internal data. This should be abstracted in such a way that a `Node` offers a public method `clearInLinks()`, `clearOutLinks()` or even `clearAllLinks()` (even beter would be to specify this method in an interface, such as `AbstractNode`).